quint/src/static/callgraph.ts

```diff
@@ -0,0 +1,247 @@
+/**
+ * Compute the call graph of Quint definitions. Technically, it is a "uses"
+ * graph, as it also captures type aliases.
+ *
+ * @author Igor Konnov, Informal Systems, 2023
+ *
+ * Copyright (c) Informal Systems 2022-2023. All rights reserved. Licensed
+ * under the Apache 2.0.  See License.txt in the project root for license
+ * information.
+ */
+
+import { Map, Record, Set } from 'immutable'
+import type { RecordOf } from 'immutable'
+
+import { IRVisitor } from '../IRVisitor'
+import { LookupTable } from '../names/lookupTable'
+import { QuintApp, QuintDef, QuintExport, QuintImport, QuintInstance, QuintModule, QuintName } from '../quintIr'
+import { QuintConstType } from '../quintTypes'
+
+/**
+ * The call graph is simply a mapping from the caller's id
+ * to the set of callee's ids.
+ */
+export type CallGraph = Map<bigint, Set<bigint>>
+
+// The record factory that we use for immutable keys
+// in CallGraphContext.importsByName (see below).
+// This requires a bit of boilerplate, see:
+// https://immutable-js.com/docs/v4.3.0/Record/
+type ImportKeyRecordProps = { importingModuleId: bigint; importedModuleName: string }
+const importKeyRecordFactory: Record.Factory<ImportKeyRecordProps> = Record({
+  importingModuleId: -1n,
+  importedModuleName: '',
+})
+type ImportKeyRecordType = RecordOf<ImportKeyRecordProps>
+function mkImportKey(importingModuleId: bigint, importedModuleName: string) {
+  return importKeyRecordFactory({ importingModuleId, importedModuleName })
+}
+
+/**
+ * A context that is needed for computing a call graph in the presence
+ * of multiple modules, imports, and exports.
+ */
+export type CallGraphContext = {
+  /**
+   * The map from a module name to the corresponding module.
+   */
+  modulesByName: Map<string, QuintModule>
+
+  /**
+   * For every module id A and the imported module name B, keep the set of
+   * import statements that refer to B in A.
+   */
+  importsByName: Map<ImportKeyRecordType, Set<bigint>>
```

Comment:
iiuc, the `bigint` is the id of the import statement?
Maybe we can add this in the comment.